File: web/src/services/api/video-upscale.ts
import { useUserStore } from "@/stores/use-user-store";
import { apiGet, apiPostEmpty, apiPostForm } from "./request";

export type VideoUpscaleJobStatus = "queued" | "uploading" | "processing" | "downloading" | "succeeded" | "failed";
export type VideoUpscaleProviderID = "volcengine-las" | "tencent-mps";
export type TencentMPSEnhancementScene = "comic" | "live" | "restore";
export type VideoUpscaleTarget = "1080p" | "2k";
export type VideoUpscaleQualityMode = "compatible" | "balanced" | "master";
export type VideoFrameInterpolationMode = "keep" | "to25" | "to30" | "double" | "to60";
export type VideoInterpolationProcessingMode = "ultra-fast" | "fast" | "medium";

export type VideoUpscalePricingRules = {
    unitPriceCny: number;
    ruleVersion: string;
    resolutionTiers: Array<{ maxShortEdge: number | null; factor: number }>;
    frameRateTiers: Array<{ maxFrameRate: number; factor: number }>;
};

export type VideoInterpolationPricingRules = {
    unitPriceCny: number;
    ruleVersion: string;
    pixelTiers: Array<{ maxPixels: number | null; fastFactor: number; mediumFactor: number }>;
};

export type VideoUpscaleCapabilities = {
    enabled: boolean;
    provider: VideoUpscaleProviderID;
    providers: Array<{
        id: VideoUpscaleProviderID;
        name: string;
        targets: VideoUpscaleTarget[];
        enhancementScenes: TencentMPSEnhancementScene[];
        defaultScene: TencentMPSEnhancementScene | "";
        costNotice: string;
        interpolation: boolean;
    }>;
    targets: VideoUpscaleTarget[];
    maxInputBytes: number;
    cloudProcessing: true;
    pricing: VideoUpscalePricingRules;
    outputQualityModes: VideoUpscaleQualityMode[];
    defaultOutputQualityMode: VideoUpscaleQualityMode;
    preserveAudioSupported: boolean;
    frameInterpolation: {
        status: "available" | "unavailable";
        modes: VideoFrameInterpolationMode[];
        processingModes: VideoInterpolationProcessingMode[];
        defaultProcessingMode: VideoInterpolationProcessingMode;
        maxTargetFrameRate: number;
        maxSourceMultiplier: number;
        pricing: VideoInterpolationPricingRules;
    };
};

export type VideoUpscaleJob = {
    id: string; projectId: string; canvasId: string; sourceNodeId: string; sourceAssetId: string;
    provider: VideoUpscaleProviderID; runId: string; providerRequestId: string; target: VideoUpscaleTarget; enhancementScene: TencentMPSEnhancementScene | ""; tencentTemplateId: number;
    status: VideoUpscaleJobStatus; progress: number; attempt: number; processingStage: string; inputWidth: number; inputHeight: number; inputDurationSeconds: number; inputFrameRate: number; inputMimeType: string; inputBytes: number;
    outputWidth: number; outputHeight: number; outputQualityMode: VideoUpscaleQualityMode; preserveAudio: boolean; frameInterpolationMode: VideoFrameInterpolationMode; interpolationMode: VideoInterpolationProcessingMode | ""; interpolationTargetFrameRate: number; interpolationRunId: string;
    estimatedBillableMinutes: number; estimatedCostCny: number; costEstimateAvailable: boolean; pricingRuleVersion: string;
    estimatedInterpolationBillableMinutes: number; estimatedInterpolationCostCny: number; interpolationCostEstimateAvailable: boolean; interpolationPricingRuleVersion: string; estimatedTotalCostCny: number;
    resultUrl: string; resultMimeType: string; resultBytes: number; errorCode: string; errorMessage: string; cloudProcessing: true;
    createdAt: string; startedAt: string; completedAt: string; updatedAt: string;
};

export type VideoUpscaleSubmitOptions = { provider: VideoUpscaleProviderID; enhancementScene?: TencentMPSEnhancementScene; target: VideoUpscaleTarget; outputQualityMode: VideoUpscaleQualityMode; preserveAudio: boolean; frameInterpolationMode: VideoFrameInterpolationMode; interpolationMode: VideoInterpolationProcessingMode };
export type CreateVideoUpscaleJobInput = VideoUpscaleSubmitOptions & { file: Blob; filename: string; projectId: string; canvasId: string; sourceNodeId: string; sourceAssetId?: string };
const token = () => useUserStore.getState().token;

export function getVideoUpscaleCapabilities() { return apiGet<VideoUpscaleCapabilities>("/api/v1/video-upscale/capabilities", undefined, token()); }
export function createVideoUpscaleJob(input: CreateVideoUpscaleJobInput) {
    const form = new FormData();
    form.append("file", input.file, input.filename); form.append("target", input.target); form.append("projectId", input.projectId); form.append("canvasId", input.canvasId); form.append("sourceNodeId", input.sourceNodeId); form.append("sourceAssetId", input.sourceAssetId || "");
    form.append("provider", input.provider); form.append("enhancementScene", input.enhancementScene || "");
    form.append("outputQualityMode", input.outputQualityMode); form.append("preserveAudio", String(input.preserveAudio)); form.append("frameInterpolationMode", input.frameInterpolationMode); form.append("interpolationMode", input.interpolationMode);
    return apiPostForm<VideoUpscaleJob>("/api/v1/video-upscale/jobs", form, token());
}
export function getVideoUpscaleJob(jobId: string) { return apiGet<VideoUpscaleJob>(`/api/v1/video-upscale/jobs/${encodeURIComponent(jobId)}`, undefined, token()); }
export function retryVideoUpscaleJob(jobId: string) { return apiPostEmpty<VideoUpscaleJob>(`/api/v1/video-upscale/jobs/${encodeURIComponent(jobId)}/retry`, token()); }

File: model/video_upscale.go
package model

type VideoUpscaleJobStatus string

const (
	VideoUpscaleJobStatusQueued      VideoUpscaleJobStatus = "queued"
	VideoUpscaleJobStatusUploading   VideoUpscaleJobStatus = "uploading"
	VideoUpscaleJobStatusProcessing  VideoUpscaleJobStatus = "processing"
	VideoUpscaleJobStatusDownloading VideoUpscaleJobStatus = "downloading"
	VideoUpscaleJobStatusSucceeded   VideoUpscaleJobStatus = "succeeded"
	VideoUpscaleJobStatusFailed      VideoUpscaleJobStatus = "failed"
)

type VideoUpscaleJob struct {
	ID                                    string                `json:"id" gorm:"primaryKey"`
	UserID                                string                `json:"-" gorm:"index"`
	ProjectID                             string                `json:"projectId" gorm:"index"`
	CanvasID                              string                `json:"canvasId" gorm:"index"`
	SourceNodeID                          string                `json:"sourceNodeId"`
	SourceAssetID                         string                `json:"sourceAssetId"`
	Provider                              string                `json:"provider"`
	EnhancementScene                      string                `json:"enhancementScene"`
	TencentTemplateID                     int64                 `json:"tencentTemplateId"`
	TencentTemplateName                   string                `json:"tencentTemplateName"`
	CloudBucket                           string                `json:"-"`
	CloudRegion                           string                `json:"-"`
	CloudInputPrefix                      string                `json:"-"`
	CloudOutputPrefix                     string                `json:"-"`
	TencentOutputObject                   string                `json:"-" gorm:"type:text"`
	InputTOSURL                           string                `json:"-" gorm:"type:text"`
	OutputTOSPath                         string                `json:"-" gorm:"type:text"`
	RunID                                 string                `json:"runId" gorm:"index"`
	ProcessingStage                       string                `json:"processingStage"`
	InterpolationMode                     string                `json:"interpolationMode"`
	InterpolationTargetFrameRate          float64               `json:"interpolationTargetFrameRate"`
	InterpolationRunID                    string                `json:"interpolationRunId" gorm:"index"`
	UpscaleResultTOSURL                   string                `json:"-" gorm:"type:text"`
	InterpolationResultTOSURL             string                `json:"-" gorm:"type:text"`
	ProviderRequestID                     string                `json:"providerRequestId"`
	Target                                string                `json:"target"`
	Status                                VideoUpscaleJobStatus `json:"status" gorm:"index"`
	Progress                              int                   `json:"progress"`
	Attempt                               int                   `json:"attempt"`
	InputWidth                            int                   `json:"inputWidth"`
	InputHeight                           int                   `json:"inputHeight"`
	InputDurationSeconds                  float64               `json:"inputDurationSeconds"`
	InputFrameRate                        float64               `json:"inputFrameRate"`
	InputMIMEType                         string                `json:"inputMimeType"`
	InputBytes                            int64                 `json:"inputBytes"`
	InputPath                             string                `json:"-" gorm:"type:text"`
	OutputWidth                           int                   `json:"outputWidth"`
	OutputHeight                          int                   `json:"outputHeight"`
	OutputQualityMode                     string                `json:"outputQualityMode"`
	PreserveAudio                         bool                  `json:"preserveAudio"`
	FrameInterpolationMode                string                `json:"frameInterpolationMode"`
	EstimatedBillableMinutes              float64               `json:"estimatedBillableMinutes"`
	EstimatedCostCNY                      float64               `json:"estimatedCostCny"`
	PricingRuleVersion                    string                `json:"pricingRuleVersion"`
	CostEstimateAvailable                 bool                  `json:"costEstimateAvailable"`
	EstimatedInterpolationBillableMinutes float64               `json:"estimatedInterpolationBillableMinutes"`
	EstimatedInterpolationCostCNY         float64               `json:"estimatedInterpolationCostCny"`
	InterpolationCostEstimateAvailable    bool                  `json:"interpolationCostEstimateAvailable"`
	InterpolationPricingRuleVersion       string                `json:"interpolationPricingRuleVersion"`
	EstimatedTotalCostCNY                 float64               `json:"estimatedTotalCostCny"`
	ResultSourceURL                       string                `json:"-" gorm:"type:text"`
	ResultURL                             string                `json:"resultUrl" gorm:"type:text"`
	ResultMIMEType                        string                `json:"resultMimeType"`
	ResultBytes                           int64                 `json:"resultBytes"`
	ErrorCode                             string                `json:"errorCode"`
	ErrorMessage                          string                `json:"errorMessage" gorm:"type:text"`
	CloudProcessing                       bool                  `json:"cloudProcessing"`
	CreatedAt                             string                `json:"createdAt" gorm:"index"`
	StartedAt                             string                `json:"startedAt"`
	CompletedAt                           string                `json:"completedAt"`
	UpdatedAt                             string                `json:"updatedAt"`
}
